src/arch/src/aarch64/gic/dist_regs.rs

```diff
@@ -0,0 +1,210 @@
+    }
+
+    #[allow(clippy::trivially_copy_pass_by_ref)]
+    fn kvm_set_dist_reg(&self, fd: &DeviceFd, offset: u32, val: &u32) -> Result<()> {
+        let gic_dist_attr = kvm_bindings::kvm_device_attr {
+            group: kvm_bindings::KVM_DEV_ARM_VGIC_GRP_DIST_REGS,
+            attr: offset as u64,
+            addr: val as *const u32 as u64,
+            flags: 0,
+        };
+        fd.set_device_attr(&gic_dist_attr).map_err(|e| {
+            Error::DeviceAttribute(e, true, kvm_bindings::KVM_DEV_ARM_VGIC_GRP_DIST_REGS)
+        })
+    }
+
+    #[allow(clippy::trivially_copy_pass_by_ref)]
+    fn kvm_get_dist_reg(&self, fd: &DeviceFd, offset: u32, val: &u32) -> Result<()> {
+        let mut gic_dist_attr = kvm_bindings::kvm_device_attr {
+            group: kvm_bindings::KVM_DEV_ARM_VGIC_GRP_DIST_REGS,
+            attr: offset as u64,
+            addr: val as *const u32 as u64,
+            flags: 0,
+        };
+
+        fd.get_device_attr(&mut gic_dist_attr).map_err(|e| {
+            Error::DeviceAttribute(e, false, kvm_bindings::KVM_DEV_ARM_VGIC_GRP_DIST_REGS)
+        })
+    }
+}
+
+// Creates a vgic distributor register.
+macro_rules! VGIC_DIST_REG {
+    ($base:expr, $bpi:expr, $length:expr) => {
+        DistReg {
+            base: $base,
+            bpi: $bpi,
+            length: $length,
+        }
+    };
+}
+
+// List with relevant distributor registers that we will be restoring.
+// Order is taken from qemu.
+static VGIC_DIST_REGS: &[DistReg] = &[
+    VGIC_DIST_REG!(GICD_CTLR, 0, 4),
+    VGIC_DIST_REG!(GICD_STATUSR, 0, 4),
+    VGIC_DIST_REG!(GICD_ICENABLER, 1, 0),
+    VGIC_DIST_REG!(GICD_ISENABLER, 1, 0),
+    VGIC_DIST_REG!(GICD_IGROUPR, 1, 0),
+    VGIC_DIST_REG!(GICD_IROUTER, 64, 0),
+    VGIC_DIST_REG!(GICD_ICFGR, 2, 0),
+    VGIC_DIST_REG!(GICD_ICPENDR, 1, 0),
+    VGIC_DIST_REG!(GICD_ISPENDR, 1, 0),
+    VGIC_DIST_REG!(GICD_ICACTIVER, 1, 0),
+    VGIC_DIST_REG!(GICD_ISACTIVER, 1, 0),
+    VGIC_DIST_REG!(GICD_IPRIORITYR, 8, 0),
+];
+
+fn get_num_interrupts(fd: &DeviceFd) -> Result<u32> {
+    let num_irqs = 0;
+
+    let mut nr_irqs_attr = kvm_bindings::kvm_device_attr {
+        group: kvm_bindings::KVM_DEV_ARM_VGIC_GRP_NR_IRQS,
+        attr: 0,
+        addr: &num_irqs as *const u32 as u64,
+        flags: 0,
+    };
+    fd.get_device_attr(&mut nr_irqs_attr).map_err(|e| {
+        Error::DeviceAttribute(e, false, kvm_bindings::KVM_DEV_ARM_VGIC_GRP_NR_IRQS)
+    })?;
+    Ok(num_irqs)
+}
+
+/// Set distributor registers of the GIC.
+pub fn set_dist_regs(fd: &DeviceFd, state: &[u32]) -> Result<()> {
+    let mut idx = 0;
+
+    for dreg in VGIC_DIST_REGS {
+        let (mut base, end) = dreg.compute_reg_range(fd)?;
+        while base < end {
+            dreg.kvm_set_dist_reg(fd, base, &state[idx])?;
+            idx += 1;
+            base += REG_SIZE as u32;
+        }
+    }
+    Ok(())
+}
+
+/// Get distributor registers of the GIC.
+pub fn get_dist_regs(fd: &DeviceFd) -> Result<Vec<u32>> {
```

Comment:
We discussed this offline. Just dropping it here.

Saving all the registers into a `Vec` might not leave us enough information. For example if we save on a firecracker version with n IRQs and try to restore on a possible later firecracker version with m IRQs, this will lead to assigning the values to wrong registers when restoring. It would be safer to have a structure that also remembers the exact register that the value belongs to. At least it would give us the option to do a translation in the future.

Reply:
I gave it another thought and decided that doing this in this PR is pretty late in the game and I might break current working functionality or even degrade currently obtained performance. I prefer to do it in a subsequent PR after the CI is also setup to run aarch64 functionality. Opened #2234.

Reply:
I also plan in actually add an integration test for this too so that I make sure that restoring from previous versions of Firecracker works. Is it ok with you?

Reply:
Yes, sounds good.